File: vertu-core/src/commonMain/kotlin/com/vertu/edge/core/flow/FlowYamlParser.kt
package com.vertu.edge.core.flow

/** Parser and serializer for Maestro-style FlowV1 YAML syntax. */
object FlowYamlParser {
  /** Parses YAML text into [FlowV1]. */
  fun parse(yaml: String): FlowV1 {
    val lines = yaml.lines()
    var appId: String? = null
    val steps = mutableListOf<FlowCommand>()
    var inSteps = false

    for (rawLine in lines) {
      val line = rawLine.trim()
      if (line.isEmpty() || line.startsWith("#")) {
        continue
      }

      if (!inSteps) {
        if (line == "---") {
          inSteps = true
          continue
        }
        if (line.startsWith("appId:")) {
          val value = line.substringAfter(':', "").trim().unquote()
          if (value.isNotEmpty()) {
            appId = value
          }
        }
        continue
      }

      if (!line.startsWith("-")) {
        continue
      }
      val payload = line.removePrefix("-").trim()
      steps += parseStep(payload)
    }

    val resolvedAppId = appId ?: error("Flow YAML is missing required field: appId")
    return FlowV1(appId = resolvedAppId, steps = steps)
  }

  /** Serializes [FlowV1] into canonical YAML syntax for file storage and sharing. */
  fun toYaml(flow: FlowV1): String {
    val body = flow.steps.joinToString(separator = "\n") { command ->
      when (command) {
        FlowCommand.LaunchApp -> "- launchApp"
        is FlowCommand.TapOn -> "- tapOn: ${command.target.toYamlScalar()}"
        is FlowCommand.InputText -> "- inputText: \"${command.value.escape()}\""
        is FlowCommand.AssertVisible -> "- assertVisible: ${command.target.toYamlScalar()}"
        is FlowCommand.AssertNotVisible -> "- assertNotVisible: ${command.target.toYamlScalar()}"
        is FlowCommand.AssertText -> "- assertText: \"${command.target.toYamlScalarForInline()}::${command.value.escape()}\""
        is FlowCommand.SelectOption -> "- selectOption: \"${command.target.toYamlScalarForInline()}::${command.option.escape()}\""
        is FlowCommand.Scroll -> "- scroll: ${command.direction.name.lowercase()}"
        is FlowCommand.Swipe -> "- swipe: ${command.direction.name.lowercase()}"
        FlowCommand.Screenshot -> "- screenshot"
        FlowCommand.ClipboardRead -> "- clipboardRead"
        is FlowCommand.ClipboardWrite -> "- clipboardWrite: \"${command.value.escape()}\""
        is FlowCommand.WindowFocus -> "- windowFocus: \"${command.target.toYamlScalar()}\""
        FlowCommand.HideKeyboard -> "- hideKeyboard"
        is FlowCommand.WaitForAnimation -> "- waitForAnimation: ${command.timeoutMs}"
      }
    }
    return buildString {
      append("appId: ")
      append(flow.appId)
      append('\n')
      append("---")
      if (body.isNotEmpty()) {
        append('\n')
        append(body)
      }
    }
  }

  private fun parseStep(payload: String): FlowCommand {
    if (!payload.contains(':')) {
      return when (payload) {
        "launchApp" -> FlowCommand.LaunchApp
        "hideKeyboard" -> FlowCommand.HideKeyboard
        "screenshot" -> FlowCommand.Screenshot
        "clipboardRead" -> FlowCommand.ClipboardRead
        else -> error("Unsupported command: $payload")
      }
    }

    val key = payload.substringBefore(':').trim()
    val rawValue = payload.substringAfter(':').trim().unquote()

    return when (key) {
      "tapOn" -> FlowCommand.TapOn(rawValue.toTarget())
      "inputText" -> FlowCommand.InputText(rawValue)
      "assertVisible" -> FlowCommand.AssertVisible(rawValue.toTarget())
      "assertNotVisible" -> FlowCommand.AssertNotVisible(rawValue.toTarget())
      "assertText" -> {
        val parts = rawValue.split("::", limit = 2)
        val target = parts.firstOrNull()?.toTarget() ?: error("assertText target is missing")
        val value = parts.getOrNull(1) ?: error("assertText value is missing")
        FlowCommand.AssertText(target = target, value = value)
      }
      "selectOption" -> {
        val parts = rawValue.split("::", limit = 2)
        val target = parts.firstOrNull()?.toTarget() ?: error("selectOption target is missing")
        val option = parts.getOrNull(1) ?: error("selectOption option is missing")
        FlowCommand.SelectOption(target = target, option = option)
      }
      "scroll" -> FlowCommand.Scroll(direction = rawValue.toDirection())
      "swipe" -> FlowCommand.Swipe(direction = rawValue.toDirection())
      "clipboardWrite" -> FlowCommand.ClipboardWrite(value = rawValue)
      "windowFocus" -> FlowCommand.WindowFocus(target = rawValue.toWindowTarget())
      "waitForAnimation" -> FlowCommand.WaitForAnimation(timeoutMs = rawValue.toLongOrNull() ?: 600L)
      else -> error("Unsupported command: $key")
    }
  }

  private fun CommandTarget.toYamlScalar(): String {
    return when {
      text != null -> "\"${text.escape()}\""
      resourceId != null -> "\"id=${resourceId.escape()}\""
      contentDescription != null -> "\"contentDescription=${contentDescription.escape()}\""
      x != null && y != null -> "\"$x,$y\""
      else -> "\"\""
    }
  }

  private fun CommandTarget.toYamlScalarForInline(): String {
    return when {
      text != null -> text
      resourceId != null -> "id=$resourceId"
      contentDescription != null -> "contentDescription=$contentDescription"
      x != null && y != null -> "$x,$y"
      else -> ""
    }
  }

  private fun WindowTarget.toYamlScalar(): String {
    return when {
      appId != null && title != null -> "appId=${appId.escape()}|title=${title.escape()}"
      appId != null -> "appId=${appId.escape()}"
      title != null -> "title=${title.escape()}"
      else -> ""
    }
  }

  private fun String.toTarget(): CommandTarget {
    return when {
      startsWith("id=") -> CommandTarget(resourceId = removePrefix("id="))
      startsWith("contentDescription=") ->
        CommandTarget(contentDescription = removePrefix("contentDescription="))
      contains(',') && split(',').size == 2 -> {
        val parts = split(',')
        CommandTarget(x = parts[0].trim().toIntOrNull(), y = parts[1].trim().toIntOrNull())
      }
      else -> CommandTarget(text = this)
    }
  }

  private fun String.toDirection(): Direction {
    return when (uppercase()) {
      "UP" -> Direction.UP
      "DOWN" -> Direction.DOWN
      "LEFT" -> Direction.LEFT
      "RIGHT" -> Direction.RIGHT
      else -> error("Unsupported direction: $this")
    }
  }

  private fun String.toWindowTarget(): WindowTarget {
    val chunks = split('|').map { it.trim() }.filter { it.isNotEmpty() }
    var appId: String? = null
    var title: String? = null
    for (chunk in chunks) {
      when {
        chunk.startsWith("appId=") -> appId = chunk.removePrefix("appId=")
        chunk.startsWith("title=") -> title = chunk.removePrefix("title=")
      }
    }
    if (appId == null && title == null) {
      title = this
    }
    return WindowTarget(appId = appId, title = title)
  }

  private fun String.unquote(): String {
    if (length < 2) {
      return this
    }
    val startsAndEndsWithDouble = startsWith('"') && endsWith('"')
    val startsAndEndsWithSingle = startsWith('\'') && endsWith('\'')
    if (startsAndEndsWithDouble || startsAndEndsWithSingle) {
      return substring(1, length - 1)
    }
    return this
  }

  private fun String.escape(): String = replace("\\", "\\\\").replace("\"", "\\\"")
}

File: iOS/VertuEdge/Sources/VertuEdgeCore/FlowYamlParser.swift
import Foundation

/// YAML parser for Maestro-style flow syntax.
public enum FlowYamlParser {
    /// Parses flow YAML into [FlowV1].
    public static func parse(_ yaml: String) throws -> FlowV1 {
        let lines = yaml.components(separatedBy: .newlines)
        var appId: String?
        var inSteps = false
        var commands: [FlowCommand] = []

        for rawLine in lines {
            let line = rawLine.trimmingCharacters(in: .whitespacesAndNewlines)
            if line.isEmpty || line.hasPrefix("#") {
                continue
            }

            if !inSteps {
                if line == "---" {
                    inSteps = true
                    continue
                }
                if line.hasPrefix("appId:") {
                    appId = String(line.dropFirst("appId:".count)).trimmingCharacters(in: .whitespacesAndNewlines)
                }
                continue
            }

            guard line.hasPrefix("-") else { continue }
            let payload = String(line.dropFirst()).trimmingCharacters(in: .whitespacesAndNewlines)
            commands.append(try parseStep(payload))
        }

        guard let resolvedAppId = appId, !resolvedAppId.isEmpty else {
            throw NSError(domain: "VertuEdge", code: 1, userInfo: [NSLocalizedDescriptionKey: "Missing appId"])
        }

        return FlowV1(appId: resolvedAppId, steps: commands)
    }

    /// Serializes a `FlowV1` back into the canonical Maestro-style YAML format.
    ///
    /// The output is round-trip compatible with `parse(_:)`: parsing the returned
    /// string produces a structurally equivalent `FlowV1`.
    public static func toYaml(_ flow: FlowV1) -> String {
        var lines: [String] = []
        lines.append("appId: \(flow.appId)")
        lines.append("---")
        for command in flow.steps {
            lines.append(serializeCommand(command))
        }
        return lines.joined(separator: "\n")
    }

    private static func serializeCommand(_ command: FlowCommand) -> String {
        switch command {
        case .launchApp:
            return "- launchApp"
        case .tapOn(let target):
            return "- tapOn: \(targetToYamlScalar(target))"
        case .inputText(let value):
            return "- inputText: \"\(escape(value))\""
        case .assertVisible(let target):
            return "- assertVisible: \(targetToYamlScalar(target))"
        case .assertNotVisible(let target):
            return "- assertNotVisible: \(targetToYamlScalar(target))"
        case .assertText(let target, let value):
            return "- assertText: \"\(targetToYamlScalarInline(target))::\(escape(value))\""
        case .selectOption(let target, let option):
            return "- selectOption: \"\(targetToYamlScalarInline(target))::\(escape(option))\""
        case .scroll(let direction, _):
            return "- scroll: \(direction.rawValue.lowercased())"
        case .swipe(let direction, _):
            return "- swipe: \(direction.rawValue.lowercased())"
        case .screenshot:
            return "- screenshot"
        case .clipboardRead:
            return "- clipboardRead"
        case .clipboardWrite(let value):
            return "- clipboardWrite: \"\(escape(value))\""
        case .windowFocus(let target):
            return "- windowFocus: \"\(windowTargetToYamlScalar(target))\""
        case .hideKeyboard:
            return "- hideKeyboard"
        case .waitForAnimation(let timeoutMs):
            return "- waitForAnimation: \(timeoutMs)"
        }
    }

    /// Produces a quoted YAML scalar for a `CommandTarget` (for standalone commands).
    private static func targetToYamlScalar(_ target: CommandTarget) -> String {
        if let text = target.text, !text.isEmpty {
            return "\"\(escape(text))\""
        }
        if let resourceId = target.resourceId, !resourceId.isEmpty {
            return "\"id=\(escape(resourceId))\""
        }
        if let desc = target.contentDescription, !desc.isEmpty {
            return "\"contentDescription=\(escape(desc))\""
        }
        if let x = target.x, let y = target.y {
            return "\"\(x),\(y)\""
        }
        return "\"\""
    }

    /// Produces an unquoted inline scalar for `CommandTarget` (used inside `::` compound values).
    private static func targetToYamlScalarInline(_ target: CommandTarget) -> String {
        if let text = target.text, !text.isEmpty { return text }
        if let resourceId = target.resourceId, !resourceId.isEmpty { return "id=\(resourceId)" }
        if let desc = target.contentDescription, !desc.isEmpty { return "contentDescription=\(desc)" }
        if let x = target.x, let y = target.y { return "\(x),\(y)" }
        return ""
    }

    private static func windowTargetToYamlScalar(_ target: WindowTarget) -> String {
        switch (target.appId, target.title) {
        case let (appId?, title?): return "appId=\(escape(appId))|title=\(escape(title))"
        case let (appId?, nil):    return "appId=\(escape(appId))"
        case let (nil, title?):    return "title=\(escape(title))"
        case (nil, nil):           return ""
        }
    }

    /// Escapes backslashes and double-quotes for YAML double-quoted strings.
    private static func escape(_ s: String) -> String {
        s.replacingOccurrences(of: "\\", with: "\\\\")
         .replacingOccurrences(of: "\"", with: "\\\"")
    }

    private static func parseStep(_ payload: String) throws -> FlowCommand {
        if !payload.contains(":") {
            if payload == "launchApp" { return .launchApp }
            if payload == "hideKeyboard" { return .hideKeyboard }
            if payload == "screenshot" { return .screenshot }
            if payload == "clipboardRead" { return .clipboardRead }
            throw NSError(domain: "VertuEdge", code: 2, userInfo: [NSLocalizedDescriptionKey: "Unsupported step: \(payload)"])
        }

        let parts = payload.split(separator: ":", maxSplits: 1).map(String.init)
        let key = parts[0].trimmingCharacters(in: .whitespacesAndNewlines)
        let value = unescape(unquote(parts[1].trimmingCharacters(in: .whitespacesAndNewlines)))

        if key == "tapOn" { return .tapOn(target: parseTarget(value)) }
        if key == "inputText" { return .inputText(value: value) }
        if key == "assertVisible" { return .assertVisible(target: parseTarget(value)) }
        if key == "assertNotVisible" { return .assertNotVisible(target: parseTarget(value)) }
        if key == "assertText" {
            let chunks = value.components(separatedBy: "::")
            guard let targetStr = chunks.first, !targetStr.isEmpty, chunks.count > 1 else {
                throw NSError(domain: "VertuEdge", code: 5, userInfo: [NSLocalizedDescriptionKey: "assertText requires target::value"])
            }
            return .assertText(target: parseTarget(targetStr), value: chunks[1...].joined(separator: "::"))
        }
        if key == "selectOption" {
            let chunks = value.components(separatedBy: "::")
            guard let targetStr = chunks.first, !targetStr.isEmpty, chunks.count > 1 else {
                throw NSError(domain: "VertuEdge", code: 6, userInfo: [NSLocalizedDescriptionKey: "selectOption requires target::option"])
            }
            return .selectOption(target: parseTarget(targetStr), option: chunks[1...].joined(separator: "::"))
        }
        if key == "scroll" { return .scroll(direction: try parseDirection(value), steps: 32) }
        if key == "swipe" { return .swipe(direction: try parseDirection(value), distanceFraction: 0.7) }
        if key == "clipboardWrite" { return .clipboardWrite(value: value) }
        if key == "windowFocus" {
            return .windowFocus(target: parseWindowTarget(value))
        }
        if key == "waitForAnimation" { return .waitForAnimation(timeoutMs: Int(value) ?? 600) }

        throw NSError(domain: "VertuEdge", code: 3, userInfo: [NSLocalizedDescriptionKey: "Unsupported command: \(key)"])
    }

    /// Removes a single wrapping pair of double- or single-quotes from a string.
    /// Unlike `replacingOccurrences`, this only strips the outermost quote characters,
    /// preserving any internal quotes in the value.
    private static func unquote(_ s: String) -> String {
        let trimmed = s.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.count >= 2 else { return trimmed }
        if (trimmed.hasPrefix("\"") && trimmed.hasSuffix("\"")) ||
           (trimmed.hasPrefix("'") && trimmed.hasSuffix("'")) {
            return String(trimmed.dropFirst().dropLast())
        }
        return trimmed
    }

    /// Unescapes YAML double-quoted string escape sequences: `\"` → `"` and `\\` → `\`.
    /// Must be called after `unquote` once the outer quotes have been stripped.
    private static func unescape(_ s: String) -> String {
        // Process character-by-character to handle sequences correctly.
        var result = ""
        result.reserveCapacity(s.count)
        var index = s.startIndex
        while index < s.endIndex {
            let ch = s[index]
            if ch == "\\" {
                let next = s.index(after: index)
                if next < s.endIndex {
                    switch s[next] {
                    case "\"": result.append("\""); index = s.index(after: next)
                    case "\\": result.append("\\"); index = s.index(after: next)
                    case "n":  result.append("\n"); index = s.index(after: next)
                    case "t":  result.append("\t"); index = s.index(after: next)
                    default:
                        // Unknown escape: preserve the backslash and continue
                        result.append(ch)
                        index = s.index(after: index)
                    }
                } else {
                    result.append(ch)
                    index = s.index(after: index)
                }
            } else {
                result.append(ch)
                index = s.index(after: index)
            }
        }
        return result
    }

    /// Parses a YAML scalar value into a typed `CommandTarget`, mirroring KMP's `toTarget()`.
    ///
    /// Recognised prefixes (produced by `targetToYamlScalar`):
    /// - `id=<value>`                → `resourceId`
    /// - `contentDescription=<value>` → `contentDescription`
    /// - `<x>,<y>` (two integers)   → `x` / `y`
    /// - anything else              → `text`
    private static func parseTarget(_ value: String) -> CommandTarget {
        if value.hasPrefix("id=") {
            return CommandTarget(resourceId: String(value.dropFirst("id=".count)))
        }
        if value.hasPrefix("contentDescription=") {
            return CommandTarget(contentDescription: String(value.dropFirst("contentDescription=".count)))
        }
        let parts = value.split(separator: ",", maxSplits: 1).map(String.init)
        if parts.count == 2,
           let x = Int(parts[0].trimmingCharacters(in: .whitespaces)),
           let y = Int(parts[1].trimmingCharacters(in: .whitespaces)) {
            return CommandTarget(x: x, y: y)
        }
        return CommandTarget(text: value)
    }

    private static func parseDirection(_ value: String) throws -> Direction {
        let upper = value.uppercased()
        if let direction = Direction(rawValue: upper) {
            return direction
        }
        throw NSError(domain: "VertuEdge", code: 4, userInfo: [NSLocalizedDescriptionKey: "Unsupported direction: \(value)"])
    }

    private static func parseWindowTarget(_ value: String) -> WindowTarget {
        let tokens = value.split(separator: "|").map { String($0).trimmingCharacters(in: .whitespacesAndNewlines) }
        var appId: String?
        var title: String?
        for token in tokens {
            if token.hasPrefix("appId=") {
                appId = String(token.dropFirst("appId=".count))
                continue
            }
            if token.hasPrefix("title=") {
                title = String(token.dropFirst("title=".count))
            }
        }

        if appId == nil && title == nil {
            title = value
        }

        return WindowTarget(appId: appId, title: title)
    }
}
